server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -863,4 +876,66 @@ public void testHistoryRetention() throws Exception {
         assertThat(recoveryStates.get(0).getIndex().totalFileCount(), is(0));
         assertThat(recoveryStates.get(0).getTranslog().recoveredOperations(), greaterThan(0));
     }
+
+    public void testDoNotInfinitelyWaitForMapping() {
```

Comment:
This test triggers the assertion failure and then tests the behavior without the assertion. I wonder if it was more appropriate to either:

1. fail harder on the assertion failure when assertions are not enabled, i.e., not throw a `MappingException` by wrapping it) and then only handle `MappingException` in `PeerTargetRecoveryService`.
2. allow MappingExceptions that come out of translog indexing (which should normally not happen?) since that can apparently happen.
3. check that we have the right mapping version up front and fail hard on any mapping failures during translog indexing.

I am not too fond of disabling assertions to test behaviour without them. I think assertions should be thought of as set in stone and unbreakable so that we do not have to worry about what happens when assertions are broken. On the other hand, I recognize that this does directly address the specific situation encountered. Let me know your thoughts on this.